guide/src/class.md

```diff
@@ -15,57 +15,111 @@ struct MyClass {
+The above example generates implementations for `PyTypeInfo`, `PyTypeObject`
+and `PyClass` for `MyClass`.
 
-## Get Python objects from `pyclass`
+Specifically, the following implementation is generated.
+
+```rust
+use pyo3::prelude::*;
+use pyo3::{PyClassShell, PyTypeInfo};
+
+struct MyClass {
+    num: i32,
+    debug: bool,
+}
+
+impl pyo3::pyclass::PyClassAlloc for MyClass {}
+
+impl PyTypeInfo for MyClass {
+    type Type = MyClass;
+    type BaseType = pyo3::types::PyAny;
+    type ConcreteLayout = PyClassShell<Self>;
+
+    const NAME: &'static str = "MyClass";
+    const MODULE: Option<&'static str> = None;
+    const DESCRIPTION: &'static str = "This is a demo class";
+    const FLAGS: usize = 0;
+
+    #[inline]
+    unsafe fn type_object() -> &'static mut pyo3::ffi::PyTypeObject {
+        static mut TYPE_OBJECT: pyo3::ffi::PyTypeObject = pyo3::ffi::PyTypeObject_INIT;
+        &mut TYPE_OBJECT
+    }
+}
 
-You can use `pyclass`es like normal rust structs.
+impl pyo3::pyclass::PyClass for MyClass {
+    type Dict = pyo3::pyclass_slots::PyClassDummySlot;
+    type WeakRef = pyo3::pyclass_slots::PyClassDummySlot;
+}
 
-However, if instantiated normally, you can't treat `pyclass`es as Python objects.
+impl pyo3::IntoPy<PyObject> for MyClass {
+    fn into_py(self, py: pyo3::Python) -> pyo3::PyObject {
+        pyo3::IntoPy::into_py(pyo3::Py::new(py, self).unwrap(), py)
+    }
+}
 
-To get a Python object which includes `pyclass`, we have to use some special methods.
+pub struct MyClassGeneratedPyo3Inventory {
+    methods: &'static [pyo3::class::PyMethodDefType],
+}
 
-### `PyRef`
+impl pyo3::class::methods::PyMethodsInventory for MyClassGeneratedPyo3Inventory {
+    fn new(methods: &'static [pyo3::class::PyMethodDefType]) -> Self {
+        Self { methods }
+    }
 
-`PyRef` is a special reference, which ensures that the referred struct is a part of
-a Python object, and you are also holding the GIL.
+    fn get_methods(&self) -> &'static [pyo3::class::PyMethodDefType] {
+        self.methods
+    }
+}
+
+impl pyo3::class::methods::PyMethodsInventoryDispatch for MyClass {
+    type InventoryType = MyClassGeneratedPyo3Inventory;
+}
+
+pyo3::inventory::collect!(MyClassGeneratedPyo3Inventory);
+
+# let gil = Python::acquire_gil();
+# let py = gil.python();
+# let cls = py.get_type::<MyClass>();
+# pyo3::py_run!(py, cls, "assert cls.__name__ == 'MyClass'")
+```
 
-You can get an instance of `PyRef` by `PyRef::new`, which does 3 things:
-1. Allocates a Python object in the Python heap
-2. Copies the Rust struct into the Python object
-3. Returns a reference to it
 
-You can use `PyRef` just like `&T`, because it implements `Deref<Target=T>`.
+## Get Python objects from `pyclass`
+You sometimes need to convert your `pyclass` into a Python object in Rust code (e.g., for testing it).
+
+You can use `PyClassShell<T>` or `Py<T>` for this purpose.
+
+### `PyClassShell`
+
+`PyClassShell` is a special type made for `PyClass`.
```

Comment:
```suggestion
`PyClassShell` is a special type made for `pyclass`.
```
I'm not sure if this sentence helps, maybe we could drop it? I found the documentation for `PyRef` very clear, does it still apply for `PyClassShell`?

> 
> You can get an instance of `PyRef` by `PyRef::new`, which does 3 things:	
> 1. Allocates a Python object in the Python heap	
> 2. Copies the Rust struct into the Python object	
> 3. Returns a reference to it	
>

Reply:
Hmm, 'Copies the Rust struct into the Python object' is not correct for pyclass... :thinking: 
But I try to enhance it.